src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java

```diff
@@ -104,35 +111,90 @@ public List<LocalCIBuildJobQueueItem> getProcessingJobsForCourse(long courseId)
     }
 
-    /**
-     * Get first build job item from the queue. If it exists, process build job and after completion,
-     * try to process next item.
-     */
-    private void processBuild() {
+    private ProgrammingExerciseParticipation retrieveParticipationWithRetry(Long participationId) {
+        int maxRetries = 5;
+        int retries = 0;
+        while (retries < maxRetries) {
+            try {
+                return (ProgrammingExerciseParticipation) participationRepository.findByIdElseThrow(participationId);
+            }
+            catch (Exception e) {
+                log.debug("Error while retrieving participation with id " + participationId + " from database: " + e.getMessage());
+                log.info("Retrying to retrieve participation with id " + participationId + " from database");
+                retries++;
+                try {
+                    Thread.sleep(1000);
+                }
+                catch (InterruptedException e1) {
+                    log.error("Error while waiting for participation with id " + participationId + " to be persisted in database: " + e1.getMessage());
+                }
+            }
+        }
+        throw new IllegalStateException("Could not retrieve participation with id " + participationId + " from database after " + maxRetries + " retries");
+    }
+
+    private void checkAvailabilityAndProcessNextBuild() {
+        lock.lock();
+
+        if (!nodeIsAvailable()) {
+            log.info("Node has no available threads currently");
+            lock.unlock();
+            return;
+        }
 
         if (queue.isEmpty()) {
+            lock.unlock();
             return;
         }
         // need to add the build job to processingJobs before taking it from the queue,
         // so it can be later added back to the queue if the node fails
         LocalCIBuildJobQueueItem buildJob;
 
         // lock the queue to prevent multiple nodes from processing the same build job
-        lock.lock();
+        fLock.lock();
         try {
             buildJob = addToProcessingJobs();
         }
         finally {
-            lock.unlock();
+            fLock.unlock();
         }
 
+        lock.unlock();
+        processBuild(buildJob);
+    }
+
+    private LocalCIBuildJobQueueItem addToProcessingJobs() {
+        LocalCIBuildJobQueueItem buildJob = queue.poll();
+        if (buildJob != null) {
+            Long participationId = buildJob.getParticipationId();
+            buildJob.setBuildStartDate(System.currentTimeMillis());
+            buildJob.setExpirationTime(System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(180));
+            processingJobs.put(participationId, buildJob);
+            localProcessingJobs++;
+        }
+        return buildJob;
+    }
+
+    // Checks whether the node has at least one thread available for a new build job
+    private Boolean nodeIsAvailable() {
```

Comment:
```suggestion
    private boolean nodeIsAvailable() {
```
non-nullable primitive type should be used when possible